app/actions/websocket/categories.ts

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {fetchMyCategories, receivedCategoryOrder} from '@mm-redux/actions/channel_categories';
+import {getCurrentTeamId} from '@mm-redux/selectors/entities/teams';
+import {ActionResult, DispatchFunc, GetStateFunc} from '@mm-redux/types/actions';
+import {WebSocketMessage} from '@mm-redux/types/websocket';
+
+export function handleSidebarCategoryCreated(msg: WebSocketMessage) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc): Promise<ActionResult> => {
+        const state = getState();
+        const currentTeamId = getCurrentTeamId(state);
+
+        if (msg.broadcast.team_id !== currentTeamId) {
+            // The new category will be loaded when we switch teams.
+            return {data: false};
+        }
+
+        // Fetch all categories, including ones that weren't explicitly updated, in case any other categories had channels
+        // moved out of them.
+        dispatch(fetchMyCategories(msg.broadcast.team_id));
+
+        return {data: true};
+    };
+}
+
+export function handleSidebarCategoryUpdated(msg: WebSocketMessage) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc): Promise<ActionResult> => {
+        const state = getState();
+
+        if (msg.broadcast.team_id !== getCurrentTeamId(state)) {
+            // The updated categories will be loaded when we switch teams.
+            return {data: false};
+        }
+
+        // Fetch all categories in case any other categories had channels moved out of them.
+        dispatch(fetchMyCategories(msg.broadcast.team_id));
+
+        return {data: true};
+    };
+}
+
+export function handleSidebarCategoryDeleted(msg: WebSocketMessage) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc): Promise<ActionResult> => {
+        const state = getState();
+
+        if (msg.broadcast.team_id !== getCurrentTeamId(state)) {
+            // The category will be removed when we switch teams.
+            return {data: false};
+        }
+
+        // Fetch all categories since any channels that were in the deleted category were moved to other categories.
+        dispatch(fetchMyCategories(msg.broadcast.team_id));
+
+        return {data: true};
+    };
+}
+
```

Comment:
Don't `handleSidebarCategoryCreated`, `handleSidebarCategoryUpdated`, and `handleSidebarCategoryDeleted` all have the same implementation?

Reply:
Agree 100%.

Also can there be further improvements on the WS event so there is no need to fetch everything even when the category is collapsed/expanded? Or that does not trigger any of this events?

Reply:
Also all of this should check for legacy cause as I user my setting could be set to use the legacy sidebar, right?